pageserver/src/pgdatadir_mapping.rs

```diff
@@ -324,83 +340,80 @@ impl Timeline {
 
         let mut found_smaller = false;
         let mut found_larger = false;
+
+        // This is a search budget to not make the search too expensive
+        let mut budget = 1000u16;
         while low < high {
             // cannot overflow, high and low are both smaller than u64::MAX / 2
-            let mid = (high + low) / 2;
-
-            let cmp = self
-                .is_latest_commit_timestamp_ge_than(
-                    search_timestamp,
-                    Lsn(mid * 8),
-                    &mut found_smaller,
-                    &mut found_larger,
-                    ctx,
-                )
-                .await?;
+            // this always holds: low <= mid_start < high
+            let mid_start = (high + low) / 2;
+            let mut mid = mid_start;
+
+            let mut max = None;
+
+            // Linear search for an lsn that has a commit timestamp.
```

Comment:
I have something planned that does an exhausting search that can deal with arbitrary large blocks of `get_max_timestamp_for_lsn` returning `None` (aka huge commits). I would iterate over the numbers from mid to high in a way that is all of exhaustive (so visits all numbers, giving us the proof that we need to update `high` if the search fails for all values between mid and high), injective (so visits them all exactly once), but also exponential.

The worst-case is always going to be that we have to search every single possible lsn value (in order to show that all lsns yield None you have to try all lsns), but the average case, where not everything is None, can be improved to not be linear.

Reply:
the search stuff is implemented now (`ExpCounter`).